crates/wasmtime/src/store.rs

```diff
@@ -1052,10 +1088,33 @@ impl<T> StoreInner<T> {
     }
 
     pub fn call_hook(&mut self, s: CallHook) -> Result<(), Trap> {
-        if let Some(hook) = &mut self.call_hook {
-            hook(&mut self.data, s)
+        // Need to borrow async_cx before the mut borrow of the limiter.
+        // self.async_cx() panicks when used with a non-async store, so
+        // wrap this in an option.
+        #[cfg(feature = "async")]
+        let async_cx = if self.inner.async_support() {
+            Some(self.inner.async_cx())
         } else {
-            Ok(())
+            None
+        };
+
+        match self.call_hook {
+            Some(CallHookInner::Sync(ref mut hook)) => hook(&mut self.data, s),
```

Comment:
FWIW if you so desire you can get rid of the pesky `ref mut` on `ref mut hook` by having the match expression be `&mut self.call_hook`